model-serving/content.md

```diff
@@ -0,0 +1,292 @@
+
+
+##  Install the required packages
+
+Run ```pip install mlflow[extras]==1.2.0 "mflux-ai>=0.3.0"```  in your terminal.
+
+
+## Tutorial
+
+In this tutorial we will deploy a machine learning model as a REST API using two approaches:
+
+* Using Flask
+* Using MLFlow's build in functionality.
+
+In both approaches we will deploy it using  a [Docker](https://www.docker.com/) container.
+
+### Train and save a model
+
+We train and save a decision tree classifier on the [Iris](https://en.wikipedia.org/wiki/Iris_flower_data_set) data set.
+The model wil be stored in MFlux.ai.
+Create a file train_model.py and paste the following code:
+
+```python
+import mflux_ai
+import mlflow.sklearn
+import os
+import pickle
+from sklearn import datasets
+from sklearn.model_selection import train_test_split
+from sklearn.tree import DecisionTreeClassifier
+
+iris = datasets.load_iris()
+input_data = iris["data"]
+target_data = iris["target"]
+model = DecisionTreeClassifier()
+model.fit(input_data, target_data)
+
+# Note: in the following line, insert the project token shown on your dashboard page.
+mflux_ai.init("your_project_token_goes_here")
+mlflow.sklearn.log_model(model, "model")
+
+```
+
+Now, run this file: ```python train_model.py```.
+
+### REST API with Flask
+
+Create a file app.py inside a folder named ```app``` and paste the following code:
+
+```python
+import mflux_ai
+import mlflow.sklearn
+import numpy as np
+import pickle
+from flask import Flask, request, send_from_directory
+from flask import jsonify
+
+app = Flask(__name__, static_folder="app")
+
+# Note: in the following line, insert the project token shown on your dashboard page.
+mflux_ai.init("your_project_token_goes_here")
+
+model = mlflow.sklearn.load_model(
+    "s3://mlflow/0/RUN_ID_GOES_HERE/artifacts/model"
+)
+
+
+@app.route("/predict/", methods=['POST'])
+def predict():
+    response = {"success": False}
+    if request.method == "POST":
+        if request.data:
+            data = request.json.get('data')
+            predictions = model.predict(np.asarray(data))
+            response["success"] = True
+            response["predictions"] = predictions.tolist()
+
+    return jsonify(response)
+
+```
+
+In ```app.py```, replace ```RUN_ID_GOES_HERE``` with the actual run id that you found in the model tracking UI.
+
+Let's go through the code. The first code snippet imports the packages and initalizes the Flask application
+
+
+```python
+import mflux_ai
+import mlflow.sklearn
+import numpy as np
+import pickle
+from flask import Flask, request, send_from_directory
+from flask import jsonify
+
+app = Flask(__name__, static_folder="app")
+```
+
+
+Next, we load our trained model from MFlux.ai.
+```python
+# Note: in the following line, insert the project token shown on your dashboard page.
+mflux_ai.init("your_project_token_goes_here")
+
+model = mlflow.sklearn.load_model(
+    "s3://mlflow/0/RUN_ID_GOES_HERE/artifacts/model"
```

Comment:
Note: Ideally, one might want to bake the model binary in the built docker image, so we might want to look into that later. That approach will reduce a potential source of errors and significantly speed up server startup